NFT/src/main.sw

```diff
@@ -0,0 +1,359 @@
+    ///
+    /// # Reverts
+    ///
+    /// * When the constructor function has already been called.
+    /// * When the `token_supply` is set to 0.
+    /// * When `access_control` is set to true and no admin was given.
+    #[storage(read, write)]fn constructor(admin: Option<Identity>, access_control: bool, token_supply: u64) {
+        // This function can only be called once so if the token supply is already set it has
+        // already been called
+        require(storage.token_supply == 0, InitError::CannotReinitialize);
+        // The number of tokens that can be minted cannot be 0
+        require(token_supply != 0, InputError::TokenSupplyCannotBeZero);
+        // Access control is set to true but there was no admin given
+        require((access_control && admin.is_some()) || admin.is_none(), InitError::AccessControlSetAndAdminIsNone);
+
+        // Store the information given
+        storage.access_control = access_control;
+        storage.admin = admin;
+        storage.token_supply = token_supply;
+    }
+
+    /// Mints a specified amount of tokens to the given `to` `Identity`. Once a token has been minted,
+    /// it can be transfered and burned. Calling this mint function will increment the `total_count`.
+    /// If the NFT contract has not yet been initalized, any attempts to mint will fail as the
+    /// `total_supply` has not yet been set.
+    ///
+    /// # Arguments
+    ///
+    /// * `to` - The `Identity` which will own the minted tokens.
+    /// * `amount` - The `u64` number of tokens to be minted in this transaction.
+    ///
+    /// # Reverts
+    ///
+    /// * When the sender attempts to mint more tokens than total supply.
+    /// * When the sender is not the admin and `access_control` is set.
+    #[storage(read, write)]fn mint(to: Identity, amount: u64) {
+        // The current number of tokens minted plus the amount to be minted cannot be
+        // greater than the total supply
+        require(storage.token_supply >= (storage.token_count + amount), InputError::NotEnoughTokensToMint);
+
+        // Ensure that the sender is the admin if this is a controlled access mint
+        require(!storage.access_control || (storage.admin.is_some() && msg_sender().unwrap() == storage.admin.unwrap()), AccessError::SenderDoesNotHaveAccessControl);
+
+        // Mint as many tokens as the sender has asked for
+        let mut index = 0;
+        let mut minted_tokens = ~Vec::new::<u64>();
+        while index < amount {
+            // Increment the token count
+            storage.token_count = storage.token_count + 1;
+
+            // Create the metadata for this new token with the owner
+            let meta_data = ~MetaData::new(Option::None(), to);
+            storage.meta_data.insert(storage.token_count, Option::Some(meta_data));
+
+            // Increase the balance of the new owner
+            storage.balances.insert(to, storage.balances.get(to) + 1);
```

Comment:
I would take this out of the loop so you're not writing to storage more than once.
Just do `storage.balances.insert(to, storage.balances.get(to) + amount);`

Reply:
I like this much better